src/runtime/waiter.rs

```diff
@@ -1,24 +1,21 @@
 use crate::*;
 
-/// A simple blocking waiter used by the generated bindings and should not generally be used directly.
+/// A simple blocking waiter used by the generated bindings and should not be used directly.
 pub struct Waiter(RawPtr);
+pub struct WaiterSignaler(RawPtr);
```

Comment:
This can made much safer if you tie the lifetime of `WaiterSignaler` was tied to the lifetime of `Waiter`. You can do this by having a `WaitSignaler<'a>` instead which you get by calling `signaler(&'a self)` on `Waiter`

Reply:
nice!

Reply:
Doesn't quite work:

```
error[E0597]: `waiter` does not live long enough
   --> D:\git\winrt-rs\target\debug\build\winrt-4b6f2188f7967158\out\winrt.rs:298:36
    |
298 |                     let signaler = waiter.signaler();
    |                                    ^^^^^^-----------
    |                                    |
    |                                    borrowed value does not live long enough
    |                                    argument requires that `waiter` is borrowed for `'static`
...
306 |                 }
    |                 - `waiter` dropped here while still borrowed
```

Reply:
The compiler doesn't know that the completed handler's lifetime is bounded by the enclosing function.

Reply:
Ah yes that's true. 😦 I guess we have to go back. We should just document the unsafe method then

Reply:
Added a comment.

Reply:
For reference, here's the implementation that would have been safe.

```rust
pub struct Waiter(RawPtr);
pub struct WaiterSignaler<'a>(RawPtr, std::marker::PhantomData<&'a Waiter>);

impl Waiter {
    pub fn new() -> Waiter {
        unsafe {
            Self(CreateEventW(
                std::ptr::null_mut(),
                1,
                0,
                std::ptr::null_mut(),
            ))
        }
    }

    pub fn signaler<'a>(&'a self) -> WaiterSignaler<'a> {
        WaiterSignaler(self.0, std::marker::PhantomData)
    }
}

impl<'a> WaiterSignaler<'a> {
    pub fn signal(&self) {
        unsafe {
            SetEvent(self.0);
        }
    }
}

impl Drop for Waiter {
    fn drop(&mut self) {
        unsafe {
            WaitForSingleObject(self.0, 0xFFFFFFFF);
            CloseHandle(self.0);
        }
    }
}
```